chromium_src/components/keyed_service/core/dependency_manager.cc

```diff
@@ -0,0 +1,21 @@
+/* Copyright (c) 2024 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#include "components/keyed_service/core/dependency_manager.h"
+
+#define DisallowKeyedServiceFactoryRegistration \
+  DisallowKeyedServiceFactoryRegistration_ChromiumImpl
+
+#include "src/components/keyed_service/core/dependency_manager.cc"
+
+#undef DisallowKeyedServiceFactoryRegistration
+
+void DependencyManager::DisallowKeyedServiceFactoryRegistration(
+    const std::string& registration_function_name_error_message) {
+#if !BUILDFLAG(IS_IOS)
+  DisallowKeyedServiceFactoryRegistration_ChromiumImpl(
```

Comment:
do we have an issue to fix this? All factories should be created before using. I don't see why this should be disabled on iOS forever.

Reply:
The main thing is that iOS overrides `EnsureBrowserStateKeyedServiceFactoriesBuilt` and doesn't call the parent method because we dont actually _want_ to register all of Chrome iOS's factories since a vast majority are not used. We can certainly remove this override and then begin the process of adding in each factory that is called but we'd need to audit Chromium's `EnsureBrowserStateKeyedServiceFactoriesBuilt` method on a regular basis to ensure we add anything that may be used implicitly by us. It will also not actually crash on launch with some factories since it would only hit on first usage of said factory which is the more worrying part.

Reply:
I'm not following, this is a generic dependency manager method and overriding `EnsureBrowserStateKeyedServiceFactoriesBuilt` shouldn't have any impact on this. This method won't cause any factories to be initialized that aren't explicitly being called so I think we're probably overriding the wrong thing here to fix whatever new issue you're seeing

Reply:
We're overriding this because it was introduced in one of `IOSChromeMainParts`s calls (https://source.chromium.org/chromium/chromium/src/+/main:ios/chrome/browser/web/model/chrome_main_parts.mm;l=316), which we already subclass but it would require also re-implementing the rest of the method call.

Reply:
discussed with @kylehickinson and there will be a follow-up to make sure the factories that getting initialized anyway get registered when they are supposed to be

Reply:
Opened issue here: https://github.com/brave/brave-browser/issues/39222